### ISSUE TITLE
correct way to leverage pre-existing, external modules?

### ISSUE THREAD

**Issue by koomie** [spack/spack#9083]
##### Background

Over in openhpc land we were recently talking about possibly providing users a mechanism to register ohpc-style packages they may have installed with spack to avoid rebuilding items already present.  I've tried to do this using the `modules` capability mentioned in the docs (https://spack.readthedocs.io/en/latest/build_settings.html), but seem to missing something. Not sure if it's a bug or if I just don't have the config right, but submitting per request and will highlight below what I tried.

Below is an example config for `packages.yaml` where I am trying to have spack know about the openhpc `autotools` module which provides access to autoconf, automake, and libtool binaries via standard PATH settings.

```
 autoconf:
    modules:
       autoconf@2.69: autotools
       buildable: False
```

I found a spec package in spack (libconfig) which has a dependency on these tools, so I have been using it as a quick litmus test. When I try to build the package, it builds a few dependencies and then errors off on the external autoconf config:

```
# spack -k build libconfig
==> Warning: You asked for --insecure. Will NOT check SSL certificates.
==> Checking dependencies for libconfig
==> Installing libsigsegv
==> Fetching https://ftp.gnu.org/gnu/libsigsegv/libsigsegv-2.11.tar.gz
######################################################################## 100.0%
==> Staging archive: /opt/ohpc/admin/spack/0.11.2/var/spack/stage/libsigsegv-2.11-nwbx3xloeloijlyclcpkhaffbhcv4ac2/libsigsegv-2.11.tar.gz
==> Created stage in /opt/ohpc/admin/spack/0.11.2/var/spack/stage/libsigsegv-2.11-nwbx3xloeloijlyclcpkhaffbhcv4ac2
==> No patches needed for libsigsegv
==> Building libsigsegv [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed libsigsegv
  Fetch: 5.00s.  Build: 8.87s.  Total: 13.88s.
[+] /opt/ohpc/admin/spack/0.11.2/opt/spack/linux-centos7-x86_64/gcc-7.3.0/libsigsegv-2.11-nwbx3xloeloijlyclcpkhaffbhcv4ac2
==> Installing m4
==> Fetching https://ftp.gnu.org/gnu/m4/m4-1.4.18.tar.gz
######################################################################## 100.0%
==> Staging archive: /opt/ohpc/admin/spack/0.11.2/var/spack/stage/m4-1.4.18-l5ndpki5yduibmsz7dywpopnlx2s7sxr/m4-1.4.18.tar.gz
==> Created stage in /opt/ohpc/admin/spack/0.11.2/var/spack/stage/m4-1.4.18-l5ndpki5yduibmsz7dywpopnlx2s7sxr
==> Applied patch gnulib-pgi.patch
==> Building m4 [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed m4
  Fetch: 0.67s.  Build: 45.24s.  Total: 45.92s.
[+] /opt/ohpc/admin/spack/0.11.2/opt/spack/linux-centos7-x86_64/gcc-7.3.0/m4-1.4.18-l5ndpki5yduibmsz7dywpopnlx2s7sxr
==> autoconf@2.69 : has external module in autotools
==> autoconf@2.69 : is actually installed in "/opt/ohpc/admin/spack/0.11.2/modules:/opt/ohpc/pub/moduledeps/gnu7-openmpi3:/opt/ohpc/pub/moduledeps/gnu7:/opt/ohpc/admin/modulefiles:/opt/ohpc/pub/modulefiles"
==> autoconf@2.69 : generating module file
==> Warning: Module file already exists : skipping creation
file : /opt/ohpc/admin/spack/0.11.2/share/spack/dotkit/linux-centos7-x86_64/autoconf-2.69-gcc-7.3.0-5czm2l2.dk
spec : autoconf@2.69%gcc@7.3.0 arch=linux-centos7-x86_64
==> autoconf@2.69 : registering into DB
==> libtool@2.4.6 : externally installed in /opt/ohpc/pub/utils/autotools
==> libtool@2.4.6 : generating module file
==> libtool@2.4.6 : registering into DB
==> automake@1.15 : externally installed in /opt/ohpc/pub/utils/autotools
==> automake@1.15 : generating module file
==> automake@1.15 : registering into DB
==> Installing libconfig
==> Fetching https://github.com/hyperrealm/libconfig/archive/v1.5.tar.gz
######################################################################## 100.0%
==> Staging archive: /opt/ohpc/admin/spack/0.11.2/var/spack/stage/libconfig-1.5-vfqtajwumv2zuoc7umupu7xwmk5bjeoh/v1.5.tar.gz
==> Created stage in /opt/ohpc/admin/spack/0.11.2/var/spack/stage/libconfig-1.5-vfqtajwumv2zuoc7umupu7xwmk5bjeoh
==> No patches needed for libconfig
==> Building libconfig [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Error: ProcessError: /opt/ohpc/admin/spack/0.11.2/opt/spack/"/opt/ohpc/admin/spack/0.11.2/modules:/opt/ohpc/pub/moduledeps/gnu7-openmpi3:/opt/ohpc/pub/moduledeps/gnu7:/opt/ohpc/admin/modulefiles:/opt/ohpc/pub/modulefiles"/bin/autoreconf: No such file or directory
    Command: '/opt/ohpc/admin/spack/0.11.2/opt/spack/"/opt/ohpc/admin/spack/0.11.2/modules:/opt/ohpc/pub/moduledeps/gnu7-openmpi3:/opt/ohpc/pub/moduledeps/gnu7:/opt/ohpc/admin/modulefiles:/opt/ohpc/pub/modulefiles"/bin/autoreconf' '-ivf'
See build log for details:
  /opt/ohpc/admin/spack/0.11.2/var/spack/stage/libconfig-1.5-vfqtajwumv2zuoc7umupu7xwmk5bjeoh/libconfig-1.5/spack-build.out
```

Looking at the output, it seems to me that it registers autoconf as coming from an external module, but somehow uses the path provided by MODULEPATH instead of PATH when it tries to run `bin/autoreconf`.  Interestingly, spack does call out the correct (sub)path to other binaries visible in this “autotools” module like `libtool` and `automake` in the trace output above. 

Here is the resolution of the desired binaries from the autotools module

```
# module list

Currently Loaded Modules:
  1) autotools   2) prun/1.2   3) gnu7/7.3.0   4) openmpi3/3.1.0   5) ohpc   6) spack/0.11.2

# which autoreconf autoconf automake libtool
/opt/ohpc/pub/utils/autotools/bin/autoreconf
/opt/ohpc/pub/utils/autotools/bin/autoconf
/opt/ohpc/pub/utils/autotools/bin/automake
/opt/ohpc/pub/utils/autotools/bin/libtool

```

And here is the MODULEPATH from my env which is curiously related to the path above being used to try and run autoreconf. 

```
# echo $MODULEPATH
/opt/ohpc/admin/spack/0.11.2/modules:/opt/ohpc/pub/moduledeps/gnu7-openmpi3:/opt/ohpc/pub/moduledeps/gnu7:/opt/ohpc/admin/modulefiles:/opt/ohpc/pub/modulefiles
```

Here is the module config (we use Lmod):

```
# module show autotools
--------------------------------------------------------------------------------------------------------------------
   /opt/ohpc/pub/modulefiles/autotools:
--------------------------------------------------------------------------------------------------------------------
whatis("Name: GNU Autotools ")
whatis("Version: 1.0 ")
whatis("Category: utility, developer support ")
whatis("Keywords: System, Utility ")
whatis("Description: Developer utilities ")
prepend_path("PATH","/opt/ohpc/pub/utils/autotools/bin")
prepend_path("MANPATH","/opt/ohpc/pub/utils/autotools/share/man")
help([[This module loads the autotools collection to provide recent
versions of autoconf, automake, and libtool.
 
]])

```

Thanks for any pointers/tips.

**Comment by tgamblin** [comment]
Thanks @koomie!  @scheibelp: here's Karl's issue.

**Comment by scheibelp** [comment]
Hi @koomie, I can replicate your issue. I'll be able to start (and likely finish) a fix for it tomorrow afternoon (8/24).

It appears that when `module show` is invoked in Spack's context, it adds lines that confuse the logic which parses out the package's path. An example of the extra output:

```
==> Module contents:
==> [module line] os.environ["LMOD_DEFAULT_MODULEPATH"] = "/usr/tce/modulefiles/Core:/usr/apps/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core";
==> [module line] os.environ["MODULEPATH"] = "/usr/tce/modulefiles/MPI/mvapich2/2.2:/usr/tce/modulefiles/MPI/intel/18.0.1/mvapich2/2.2:/usr/tce/modulefiles/Compiler/intel/18.0.1:/usr/tce/modulefiles/Core:/usr/apps/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core";
==> [module line] os.environ["LMOD_DEFAULT_MODULEPATH"] = "/usr/tce/modulefiles/Core:/usr/apps/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core";
==> [module line] os.environ["MODULEPATH"] = "/usr/tce/modulefiles/MPI/mvapich2/2.2:/usr/tce/modulefiles/MPI/intel/18.0.1/mvapich2/2.2:/usr/tce/modulefiles/Compiler/intel/18.0.1:/usr/tce/modulefiles/Core:/usr/apps/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core";
==> [module line] os.environ["_ModuleTable001_"] = "X01vZHVsZVRhYmxlXz17WyJhY3RpdmVTaXplIl09NCxiYXNlTXBhdGhBPXsiL3Vzci90Y2UvbW9kdWxlZmlsZXMvQ29yZSIsIi91c3IvYXBwcy9tb2R1bGVmaWxlcyIsIi91c3Ivc2hhcmUvbW9kdWxlZmlsZXMvTGludXgiLCIvdXNyL3NoYXJlL21vZHVsZWZpbGVzL0NvcmUiLCIvdXNyL3NoYXJlL2xtb2QvbG1vZC9tb2R1bGVmaWxlcy9Db3JlIix9LFsiY19yZWJ1aWxkVGltZSJdPTUzMC4zMzgxODI0NDkzNCxbImNfc2hvcnRUaW1lIl09NC40MTk0ODQ4NTM3NDQ1LGZhbWlseT17WyJjb21waWxlciJdPSJpbnRlbCIsWyJtcGkiXT0ibXZhcGljaDIiLH0saW5hY3RpdmU9e30sbVQ9e1N0ZEVudj17WyJGTiJdPSIvdXNyL3RjZS9tb2R1bGVmaWxlcy9Db3JlL1N0ZEVudi5sdWEi";
==> [module line] os.environ["_ModuleTable002_"] = "LFsiZGVmYXVsdCJdPTAsWyJmdWxsTmFtZSJdPSJTdGRFbnYiLFsibG9hZE9yZGVyIl09NCxwcm9wVD17fSxbInNob3J0Il09IlN0ZEVudiIsWyJzdGF0dXMiXT0iYWN0aXZlIix9LGludGVsPXtbIkZOIl09Ii91c3IvdGNlL21vZHVsZWZpbGVzL0NvcmUvaW50ZWwvMTguMC4xLmx1YSIsWyJkZWZhdWx0Il09MSxbImZ1bGxOYW1lIl09ImludGVsLzE4LjAuMSIsWyJsb2FkT3JkZXIiXT0xLHByb3BUPXt9LFsic2hvcnQiXT0iaW50ZWwiLFsic3RhdHVzIl09ImFjdGl2ZSIsfSxtdmFwaWNoMj17WyJGTiJdPSIvdXNyL3RjZS9tb2R1bGVmaWxlcy9Db21waWxlci9pbnRlbC8xOC4wLjEvbXZhcGljaDIvMi4yLmx1YSIsWyJkZWZhdWx0Il09MSxbImZ1bGxOYW1lIl09Im12YXBpY2gy";
==> [module line] os.environ["_ModuleTable003_"] = "LzIuMiIsWyJsb2FkT3JkZXIiXT0yLHByb3BUPXt9LFsic2hvcnQiXT0ibXZhcGljaDIiLFsic3RhdHVzIl09ImFjdGl2ZSIsfSx0ZXhsaXZlPXtbIkZOIl09Ii91c3IvdGNlL21vZHVsZWZpbGVzL0NvcmUvdGV4bGl2ZS8yMDE2Lmx1YSIsWyJkZWZhdWx0Il09MCxbImZ1bGxOYW1lIl09InRleGxpdmUvMjAxNiIsWyJsb2FkT3JkZXIiXT0zLHByb3BUPXt9LFsic2hvcnQiXT0idGV4bGl2ZSIsWyJzdGF0dXMiXT0iYWN0aXZlIix9LH0sbXBhdGhBPXsiL3Vzci90Y2UvbW9kdWxlZmlsZXMvTVBJL212YXBpY2gyLzIuMiIsIi91c3IvdGNlL21vZHVsZWZpbGVzL01QSS9pbnRlbC8xOC4wLjEvbXZhcGljaDIvMi4yIiwiL3Vzci90Y2UvbW9kdWxlZmlsZXMvQ29tcGlsZXIvaW50ZWwv";
==> [module line] os.environ["_ModuleTable004_"] = "MTguMC4xIiwiL3Vzci90Y2UvbW9kdWxlZmlsZXMvQ29yZSIsIi91c3IvYXBwcy9tb2R1bGVmaWxlcyIsIi91c3Ivc2hhcmUvbW9kdWxlZmlsZXMvTGludXgiLCIvdXNyL3NoYXJlL21vZHVsZWZpbGVzL0NvcmUiLCIvdXNyL3NoYXJlL2xtb2QvbG1vZC9tb2R1bGVmaWxlcy9Db3JlIix9LFsic3lzdGVtQmFzZU1QQVRIIl09Ii91c3IvdGNlL21vZHVsZWZpbGVzL0NvcmU6L3Vzci9hcHBzL21vZHVsZWZpbGVzOi91c3Ivc2hhcmUvbW9kdWxlZmlsZXMvTGludXg6L3Vzci9zaGFyZS9tb2R1bGVmaWxlcy9Db3JlOi91c3Ivc2hhcmUvbG1vZC9sbW9kL21vZHVsZWZpbGVzL0NvcmUiLFsidmVyc2lvbiJdPTIsfQ==";
==> [module line] os.environ["_ModuleTable_Sz_"] = "4";
==> [module line] ----------------------------------------------------------------------------
==> [module line]    /usr/tce/modulefiles/Core/cmake/3.9.2.lua:
==> [module line] ----------------------------------------------------------------------------
==> [module line] help([[CMake Version 3.9.2
==> [module line] ]])
==> [module line] whatis("Name: CMake")
==> [module line] whatis("Version: 3.9.2")
==> [module line] whatis("Category: Tools")
==> [module line] whatis("URL: https://cmake.org/")
==> [module line] prepend_path("PATH","/usr/tce/packages/cmake/cmake-3.9.2/bin")
==> [module line] prepend_path("MANPATH","/usr/tce/packages/cmake/cmake-3.9.2/share/man")
==> [module line] 
==> [module line] 
==> Extracted module path: "/usr/tce/modulefiles/Core:/usr/apps/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core"
```

The parser (in `module_cmd.py`) searches for "PATH", which is satisfied by "MODULE_PATH"; it returns the first thing that satisfies, which in this case is "LMOD_DEFAULT_MODULEPATH".

**Comment by scheibelp** [comment]
@koomie would you have the chance to confirm that https://github.com/spack/spack/pull/9100 fixes the issue for you?

**Comment by koomie** [comment]
I can confirm that #9100 does address the initial observed issue. Looking thru the code a bit though, it seems that the logic is based on scanning all of the `module show` output and parsing to grok path. This means an ill-timed mention to `PATH` in a `module-whatis` stanza can still flub up the parser and I was able to do this locally. At a minimum, I would think it might be desirable to add some regex's (or equivalent) to only consider lines which include `prepend-path`.

But that also got me wondering if perhaps there might be another way? You look to already have wrappers for loading/unloading modules so as opposed to parsing raw modulefiles directly, perhaps you could just let modules do the parsing directly (since it is the authoritative source) and interrogate the resulting environment to get the PATH.  

Here is a quick prototype I did locally to test it out to replace the `get_path_from_module` routine.  I notice that the existing function seems to strip off `/bin` if it's present, so I had to do the same here as something else must be adding it back later. Anyway, this approach seemed to work ok in my local environment.

```
def get_path_from_module(mod):
    import difflib

    # Create a modulecmd executable                                                                                                    
    modulecmd = get_module_cmd()

    # attempt to unload external module (as it may be loaded already) and cache PATH                                                   
    exec(compile(modulecmd('unload',mod,output=str,error=str),'<string>','exec'))
    PATH1=os.environ['PATH']

    # load external module and compare to cache for PATH updates                                                                       
    exec(compile(modulecmd('load',mod,output=str,error=str),'<string>','exec'))
    PATH2=os.environ['PATH']

    if PATH1 != PATH2:
        s = difflib.SequenceMatcher(None,PATH1,PATH2)
        matches = s.get_matching_blocks()
        # we expect one match (and difflib always has trailing dummy 0-length match)                                                   
        assert(len(matches) == 2)

        duplicate = PATH2[matches[0].b:matches[0].b+matches[0].size]
        remaining = PATH2.replace(duplicate,'')

        # strip trailing ':' if present                                                                                                
        if remaining.endswith(':'):
            remaining = remaining[:-1]

        # don't understand why, but strip /bin from path if present                                                                    
        remaining = remaining[:remaining.find('/bin')]

        return(remaining)
    else:
        return("")
```

**Comment by scheibelp** [comment]
> This means an ill-timed mention to PATH in a module-what is stanza can still flub up the parser and I was able to do this locally.

To be clear, does this mean you encounter a new environment-related error with https://github.com/spack/spack/pull/9100 which still prevents you from building `libconfig`? Or does this mean that you created a test module which still produces an error (but that you can build `libconfig`).

In the first case I'd say more adjusting is needed for the pattern checking in https://github.com/spack/spack/pull/9100. In the second case I would merge https://github.com/spack/spack/pull/9100, and open an issue on how module content parsing could be improved.

**Comment by koomie** [comment]
Sorry for the delay here. Yes, I can indeed build `libconfig` with the change in #9100. But, the parsing looks a bit fragile since it is parsing items (like `whatis` clauses) that don't influence items of interest (like `PATH`, `LD_LIBRARY_PATH`,etc).  That's what lead to the comment that perhaps you might just want to consider using modules evaluation directly to potentially avoid having to re-implement it's parsing logic.

**Comment by scheibelp** [comment]
I closed this as the original issue is addressed, but I opened a new issue at #9232 to address the points in https://github.com/spack/spack/issues/9083#issuecomment-416314241.

Thanks for reporting and for providing all the info!
